myia/abstract/utils.py

```diff
@@ -145,6 +150,42 @@ def build_type(self, x: AbstractType):
     return dtype.TypeType
 
 
+# A variant of built_dtype with precise function types.
+# Use for the .type attributes of nodes
+
+
+@build_type.variant
+def build_type_fn(self, x: AbstractFunction):
```

Comment:
Yes it had some issues because `build_type` is used in a number of places and most of those don't have the correct information for complete function types.